File: snippets/lesson-10-concatmap.cs
using System;
using System.Reactive.Linq;
using System.Threading.Tasks;

var files = new[] { "photo.jpg", "report.pdf", "notes.txt" };

files.ToObservable()
    .Select(file => Observable.FromAsync(async () =>
    {
        Console.WriteLine($"Uploading {file}");
        await Task.Delay(800);
        return "OK";
    }))
    .Concat()
    .Subscribe(status => Console.WriteLine($"Result: {status}"));

File: snippets/lesson-10-concatmap.ts
import { from, of } from 'rxjs';
import { concatMap, delay } from 'rxjs/operators';

const files = ['photo.jpg', 'report.pdf', 'notes.txt'];

from(files).pipe(
  concatMap(file => {
    console.log('Uploading', file);
    return of('OK').pipe(delay(800));
  }),
).subscribe(status => console.log('Result:', status));
